File: static/utils/esdl_merge.js
/**
 *  This work is based on original code developed and copyrighted by TNO 2020.
 *  Subsequent contributions are licensed to you by the developers of such code and are
 *  made available to the Project under one or several contributor license agreements.
 *
 *  This work is licensed to you under the Apache License, Version 2.0.
 *  You may obtain a copy of the license at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Contributors:
 *      TNO         - Initial implementation
 *  Manager:
 *      TNO
 */

function send_merge_esdl() {
    let esdl_source_id = document.getElementById('esdl_merge_source')
        .options[document.getElementById('esdl_merge_source').selectedIndex].value;
    let esdl_destination_id = document.getElementById('esdl_merge_destination')
        .options[document.getElementById('esdl_merge_destination').selectedIndex].value;

    document.getElementById('esdl_merge_button').style.display = 'none';
    socket.emit('esdl_merge', {esdl_source_id: esdl_source_id, esdl_destination_id: esdl_destination_id});
}

function merge_esdl_window() {
    let sidebar_ctr = sidebar.getContainer();

    sidebar_ctr.innerHTML = '<h1>Merge ESDL</h1>';
    sidebar_ctr.innerHTML += '<p><b>Note:</b> This is very experimental functionality to test the merge library. ' +
        'Properly merging references is currently \'work in progress\'</p>'

    if (esdl_list) {
        sidebar_ctr.innerHTML += '<h3>Select energysystem to merge:</h3>';

        let select = '<select id="esdl_merge_source" style="width: 300px;">';

        for (let id in esdl_list) {
            es = esdl_list[id];
            select += '<option value="' + id +
                '">' + es['title'] + '</option>';
        }
        select += '</select>';
        sidebar_ctr.innerHTML += select;

        sidebar_ctr.innerHTML += '<h3>... into this energysystem:</h3>';

        select = '<select id="esdl_merge_destination" style="width: 300px;">';
        for (let id in esdl_list) {
            es = esdl_list[id];
            select += '<option value="' + id +
                '">' + es['title'] + '</option>';
        }
        select += '</select>';
        sidebar_ctr.innerHTML += select;
        sidebar_ctr.innerHTML += '<p><button id="esdl_merge_button" onclick="sidebar.hide();send_merge_esdl();">Merge</button></p>';
    } else {
        sidebar_ctr.innerHTML += 'No ESDLs found!';
        sidebar_ctr.innerHTML += '<button id="esdl_merge_close_button" onclick="sidebar.hide();">Close</button>';
    }

    sidebar.show();
}

File: extensions/esdl_merge.py
#  This work is based on original code developed and copyrighted by TNO 2020.
#  Subsequent contributions are licensed to you by the developers of such code and are
#  made available to the Project under one or several contributor license agreements.
#
#  This work is licensed to you under the Apache License, Version 2.0.
#  You may obtain a copy of the license at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Contributors:
#      TNO         - Initial implementation
#  Manager:
#      TNO

from flask import Flask
from flask_executor import Executor
from flask_socketio import SocketIO, emit
from extensions.session_manager import get_handler
import src.log as log
from src.merge import ESDLMerger
from src.process_es_area_bld import process_energy_system

logger = log.get_logger(__name__)


class ESDLMerge:
    def __init__(self, flask_app: Flask, socket: SocketIO, executor: Executor):
        self.flask_app = flask_app
        self.socketio = socket
        self.executor = executor
        self.merger = ESDLMerger()
        self.register()

    def register(self):
        logger.info('Registering ESDL Merge extension')

        @self.socketio.on('esdl_merge', namespace='/esdl')
        def merge(message):
            with self.flask_app.app_context():
                esh = get_handler()

                esdl_source_id = message['esdl_source_id']
                esdl_destination_id = message['esdl_destination_id']

                es_source = esh.get_energy_system(es_id=esdl_source_id)
                es_destination = esh.get_energy_system(es_id=esdl_destination_id)

                logger.info("Merging EnergySystem with id {} into EnergySystem with id {}".format(esdl_source_id,
                                                                                                  esdl_destination_id))

                self.merger.merge(es_destination, es_source)
                self.executor.submit(process_energy_system, esh=esh, filename=None,
                                     force_update_es_id=esdl_destination_id)
